jsonschema-core.xml

```diff
@@ -1353,8 +1353,8 @@
                     </t>
                     <t>
                         If present, the value for this keyword MUST be a string, and MUST represent a
-                        valid <xref target="RFC3987">IRI-reference</xref>.  This IRI-reference
-                        SHOULD be normalized, and MUST resolve to an
+                        valid <xref target="RFC3987">IRI-reference</xref> with a non-empty non-fragment
+                        component.  This IRI-reference SHOULD be normalized, and MUST resolve to an
```

Comment:
I don't think this change is necessary. The two valid options are presented as an absolute-IRI or an IRI with an empty fragment. Both absolute-IRI and an IRI are by definition non-relative, so we don't need to say "non-empty". It also says it "MUST resolve to", which also indicates a non-relative result.

Reply:
@jdesrosiers: This doesn't sound right. $id can be a relative IRI. If that relative IRI is empty (apart from fragment), it resolves to the same absolute IRI as the parent schema (or the same as its retrieval IRI if it's at the root, but that part isn't a problem). Then a subschema has the same id as its parent. Forbidding the empty non-fragment portion disallows this id collision.

@handrews: The wording is awkward, "with a non-empty non-fragment component". Maybe a new sentence would be better:

> If present, the value for this keyword MUST be a string, and MUST represent a valid IRI-reference. [unchanged]
> This IRI-reference MUST NOT be empty or consist only of an empty fragment.
> or
> This IRI-reference MUST NOT have an empty non-fragment component.
>
> This IRI-reference SHOULD be normalized ...

Reply:
@jdesrosiers I'm not sure I understand.  `"$id": "foo/bar"` is valid, both before and after this change.  It is a path-only relative IRI-reference that, resolved against a base IRI (say, `https://example.com/schemas/base`) produces an absolute-IRI (`https://example.com/foo/bar`).

@notEthan Thanks, I think it like "This IRI-reference MUST NOT be empty or consist only of an empty fragment." I checked RFC 3986 and it uses the term "empty" to describe relative references in [§4.4 Same-Document Reference](https://datatracker.ietf.org/doc/html/rfc3986#section-4.4).

Reply:
I wasn't saying that `$id` can't be relative. I was talking about the result of resolving the `$id` against the base IRI. I thought this change was trying to protect against case where you don't have a baseURI and therefore the IRI can't be fully resolved. I was pointing out that the current wording already protects against that case.

> If that relative IRI is empty (apart from fragment), it resolves to the same absolute IRI as the parent schema

Oh, ok, now I understand where this change is coming from. But, I still don't think it's necessary because we talk about not allowing duplicate identifiers more generally elsewhere. IMO, it's already covered and doesn't need to be repeated here.

Reply:
> I thought this change was trying to protect against case where you don't have a baseURI and therefore the IRI can't be fully resolved. I was pointing out that the current wording already protects against that case.

Ah, I see.  Yeah, that is handled and not improved (or made worse) by this change.

> we talk about not allowing duplicate identifiers more generally elsewhere. IMO, it's already covered and doesn't need to be repeated here.

Cool- yeah, that's the main question (for one of the commits in this PR- the rest of it remains valid as it clarifies that "already covered" part that, as currently written, kinda contradicts itself in a confusing way).

I'd like to leave this open and see if we get any more opinions from other @json-schema-org/spec-team folks (or anyone else, I just can't tag the whole universe).  I don't feel strongly on this point, so if no one advocates for the change I'll strip it out and we can go with just the clean-up of the duplicate IRI stuff.

Reply:
One option might be to note that, while syntactically allowed, `""` and `"#"` effectively produce duplicate IRIs which are considered errors per §whatever:

> Note that while the empty IRI and the empty fragment-only IRI are legal values, they produce duplicate IRIs when used in an embedded resource, which is considered an error per 8.2.3.

That's probably enough to discourage the values without requiring people to detect and block them.  I'm currently leaning towards this option.

Reply:
it might be a little odd to include a MUST NOT with the intent of heading off a route to duplicate IRIs, when the language on duplicate IRIs itself isn't a MUST NOT ("implementations SHOULD raise an error condition").